File: sdk/android/src/main/java/com/reactnativehypertracksdk/common/Serialization.kt
package com.reactnativehypertracksdk.common

import android.location.Location
import com.hypertrack.sdk.android.HyperTrack
import com.hypertrack.sdk.android.Result

/**
 * Platform-independent serialization code that converts HyperTrack data types
 * to Map<String, T> or List<T> where T is any JSON-compatible type
 */
internal object Serialization {
    fun deserializeAllowMockLocation(allowMockLocation: Serialized): WrapperResult<Boolean> =
        parse(allowMockLocation) {
            it.assertValue<String>(key = KEY_TYPE, value = TYPE_ALLOW_MOCK_LOCATION)
            it
                .get<Boolean>(KEY_VALUE)
                .getOrThrow()
        }

    fun deserializeDynamicPublishableKey(args: Serialized): WrapperResult<String> =
        parse(args) {
            it.assertValue<String>(key = KEY_TYPE, value = TYPE_DYNAMIC_PUBLISHABLE_KEY)
            it
                .get<String>(KEY_VALUE)
                .getOrThrow()
        }

    fun deserializeGeotagData(map: Serialized): WrapperResult<GeotagData> =
        parse(map) {
            val data =
                it
                    .get<Serialized>(KEY_GEOTAG_DATA)
                    .getOrThrow()
            val expectedLocationData =
                it
                    .getOptional<Serialized>(KEY_GEOTAG_EXPECTED_LOCATION)
                    .getOrThrow()
            val expectedLocation =
                expectedLocationData?.let {
                    deserializeLocation(it).getOrThrow()
                }
            val orderHandleData =
                it
                    .getOptional<Serialized>(KEY_GEOTAG_ORDER_HANDLE)
                    .getOrThrow()
            val orderHandle = orderHandleData?.let { deserializeOrderHandle(it).getOrThrow() }
            val orderStatusData =
                it
                    .getOptional<Serialized>(KEY_GEOTAG_ORDER_STATUS)
                    .getOrThrow()
            val orderStatus = orderStatusData?.let { deserializeOrderStatus(it).getOrThrow() }
            GeotagData(
                data = data,
                expectedLocation = expectedLocation,
                orderHandle = orderHandle,
                orderStatus = orderStatus,
            )
        }

    fun deserializeIsAvailable(isAvailable: Serialized): WrapperResult<Boolean> =
        parse(isAvailable) {
            it.assertValue<String>(key = KEY_TYPE, value = TYPE_IS_AVAILABLE)
            it
                .get<Boolean>(KEY_VALUE)
                .getOrThrow()
        }

    fun deserializeIsTracking(isTracking: Serialized): WrapperResult<Boolean> =
        parse(isTracking) {
            it.assertValue<String>(key = KEY_TYPE, value = TYPE_IS_TRACKING)
            it
                .get<Boolean>(KEY_VALUE)
                .getOrThrow()
        }

    fun deserializeMetadata(metadata: Serialized): WrapperResult<Serialized> =
        parse(metadata) {
            it.assertValue<String>(key = KEY_TYPE, value = TYPE_METADATA)
            it
                .get<Serialized>(KEY_VALUE)
                .getOrThrow()
        }

    fun deserializeName(name: Serialized): WrapperResult<String> =
        parse(name) {
            it.assertValue<String>(key = KEY_TYPE, value = TYPE_NAME)
            it
                .get<String>(KEY_VALUE)
                .getOrThrow()
        }

    fun deserializeOrderHandle(map: Serialized): WrapperResult<String> =
        parse(map) {
            it.assertValue<String>(key = KEY_TYPE, value = TYPE_ORDER_HANDLE)
            it
                .get<String>(KEY_VALUE)
                .getOrThrow()
        }

    fun deserializeWorkerHandle(map: Serialized): WrapperResult<String> =
        parse(map) {
            it.assertValue<String>(key = KEY_TYPE, value = TYPE_WORKER_HANDLE)
            it
                .get<String>(KEY_VALUE)
                .getOrThrow()
        }

    fun serializeAllowMockLocation(allowMockLocation: Boolean): Serialized =
        mapOf(
            KEY_TYPE to TYPE_ALLOW_MOCK_LOCATION,
            KEY_VALUE to allowMockLocation,
        )

    fun serializeDeviceId(deviceId: String): Serialized =
        mapOf(
            KEY_TYPE to TYPE_DEVICE_ID,
            KEY_VALUE to deviceId,
        )

    fun serializeDynamicPublishableKey(publishableKey: String): Serialized =
        mapOf(
            KEY_TYPE to TYPE_DYNAMIC_PUBLISHABLE_KEY,
            KEY_VALUE to publishableKey,
        )

    fun serializeError(error: HyperTrack.Error): Map<String, String> =
        mapOf(
            KEY_TYPE to TYPE_ERROR,
            KEY_VALUE to
                when (error) {
                    HyperTrack.Error.BlockedFromRunning -> "blockedFromRunning"
                    HyperTrack.Error.InvalidPublishableKey -> "invalidPublishableKey"
                    HyperTrack.Error.Location.Mocked -> "location.mocked"
                    HyperTrack.Error.Location.ServicesDisabled -> "location.servicesDisabled"
                    HyperTrack.Error.Location.ServicesUnavailable -> "location.servicesUnavailable"
                    HyperTrack.Error.Location.SignalLost -> "location.signalLost"
                    HyperTrack.Error.NoExemptionFromBackgroundStartRestrictions -> "noExemptionFromBackgroundStartRestrictions"
                    HyperTrack.Error.Permissions.Location.Denied -> "permissions.location.denied"
                    HyperTrack.Error.Permissions.Location.InsufficientForBackground -> "permissions.location.insufficientForBackground"
                    HyperTrack.Error.Permissions.Location.ReducedAccuracy -> "permissions.location.reducedAccuracy"
                    HyperTrack.Error.Permissions.Notifications.Denied -> "permissions.notifications.denied"
                },
        )

    fun serializeErrors(errors: Set<HyperTrack.Error>): List<Map<String, String>> =
        errors.map {
            serializeError(it)
        }

    fun serializeFailure(failure: List<Serialized>): Serialized =
        mapOf(
            KEY_TYPE to TYPE_RESULT_FAILURE,
            KEY_VALUE to failure,
        )

    fun serializeFailure(failure: Serialized): Serialized =
        mapOf(
            KEY_TYPE to TYPE_RESULT_FAILURE,
            KEY_VALUE to failure,
        )

    fun serializeIsAvailable(isAvailable: Boolean): Serialized =
        mapOf(
            KEY_TYPE to TYPE_IS_AVAILABLE,
            KEY_VALUE to isAvailable,
        )

    fun serializeIsInsideGeofence(isInsideGeofence: Result<Boolean, HyperTrack.LocationError>): Serialized =
        when (isInsideGeofence) {
            is Result.Failure -> {
                serializeFailure(serializeLocationError(isInsideGeofence.failure))
            }

            is Result.Success -> {
                serializeSuccess(
                    mapOf(
                        KEY_TYPE to TYPE_IS_INSIDE_GEOFENCE,
                        KEY_VALUE to isInsideGeofence.success,
                    ),
                )
            }
        }

    fun serializeIsTracking(isTracking: Boolean): Serialized =
        mapOf(
            KEY_TYPE to TYPE_IS_TRACKING,
            KEY_VALUE to isTracking,
        )

    fun serializeLocateResult(locationResult: Result<HyperTrack.Location, Set<HyperTrack.Error>>): Serialized =
        when (locationResult) {
            is Result.Failure -> {
                serializeFailure(serializeErrors(locationResult.failure))
            }

            is Result.Success -> {
                serializeLocationSuccess(locationResult.success)
            }
        }

    fun serializeLocationResult(locationResult: Result<HyperTrack.Location, HyperTrack.LocationError>): Serialized =
        when (locationResult) {
            is Result.Failure -> {
                serializeLocationErrorFailure(locationResult.failure)
            }

            is Result.Success -> {
                serializeLocationSuccess(locationResult.success)
            }
        }

    fun serializeLocationErrorFailure(locationError: HyperTrack.LocationError): Serialized = serializeFailure(serializeLocationError(locationError))

    fun serializeLocationSuccess(location: HyperTrack.Location): Serialized = serializeSuccess(serializeLocation(location))

    fun serializeLocationWithDeviationSuccess(locationWithDeviation: HyperTrack.LocationWithDeviation): Serialized =
        serializeSuccess(
            serializeLocationWithDeviation(
                locationWithDeviation,
            ),
        )

    fun serializeMetadata(metadata: Serialized): Serialized =
        mapOf(
            KEY_TYPE to TYPE_METADATA,
            KEY_VALUE to metadata,
        )

    fun serializeName(name: String): Serialized =
        mapOf(
            KEY_TYPE to TYPE_NAME,
            KEY_VALUE to name,
        )

    fun serializeOrders(orders: Collection<HyperTrack.Order>): Serialized =
        mapOf(
            KEY_TYPE to TYPE_ORDERS,
            KEY_VALUE to
                orders.mapIndexed { index, order ->
                    mapOf(
                        KEY_ORDER_HANDLE to order.orderHandle,
                        KEY_ORDER_INDEX to index,
                        // beware not to call isInsideGeofence here, it's a computed property
                    )
                },
        )

    fun serializeSuccess(success: Serialized): Serialized =
        mapOf(
            KEY_TYPE to TYPE_RESULT_SUCCESS,
            KEY_VALUE to success,
        )

    fun serializeWorkerHandle(workerHandle: String): Serialized =
        mapOf(
            KEY_TYPE to TYPE_WORKER_HANDLE,
            KEY_VALUE to workerHandle,
        )

    private fun deserializeLocation(map: Serialized): WrapperResult<Location> =
        parse(map) {
            it.assertValue<String>(key = KEY_TYPE, value = TYPE_LOCATION)
            val value =
                it
                    .get<Serialized>(KEY_VALUE)
                    .getOrThrow()
            parse(value) { parser ->
                val latitude =
                    parser
                        .get<Double>(KEY_LATITUDE)
                        .getOrThrow()
                val longitude =
                    parser
                        .get<Double>(KEY_LONGITUDE)
                        .getOrThrow()
                Location("api").also {
                    it.latitude = latitude
                    it.longitude = longitude
                }
            }.getOrThrow()
        }

    private fun deserializeOrderStatus(map: Serialized): WrapperResult<HyperTrack.OrderStatus> =
        parse(map) {
            when (it.get<String>(KEY_TYPE).getOrThrow()) {
                TYPE_GEOTAG_ORDER_STATUS_CLOCK_IN -> HyperTrack.OrderStatus.ClockIn
                TYPE_GEOTAG_ORDER_STATUS_CLOCK_OUT -> HyperTrack.OrderStatus.ClockOut
                TYPE_GEOTAG_ORDER_STATUS_CUSTOM ->
                    HyperTrack.OrderStatus.Custom(
                        it.get<String>(KEY_VALUE).getOrThrow(),
                    )

                else -> throw Error("Unknown order status: $map")
            }
        }

    private fun serializeLocation(location: HyperTrack.Location): Serialized =
        mapOf(
            KEY_TYPE to TYPE_LOCATION,
            KEY_VALUE to
                mapOf(
                    KEY_LATITUDE to location.latitude,
                    KEY_LONGITUDE to location.longitude,
                ),
        )

    private fun serializeLocationWithDeviation(locationWithDeviation: HyperTrack.LocationWithDeviation): Serialized =
        mapOf(
            KEY_TYPE to TYPE_LOCATION_WITH_DEVIATION,
            KEY_VALUE to
                mapOf(
                    KEY_LOCATION to serializeLocation(locationWithDeviation.location),
                    KEY_DEVIATION to locationWithDeviation.deviation,
                ),
        )

    private fun serializeLocationError(locationError: HyperTrack.LocationError): Serialized =
        when (locationError) {
            HyperTrack.LocationError.NotRunning -> {
                mapOf(KEY_TYPE to TYPE_LOCATION_ERROR_NOT_RUNNING)
            }

            HyperTrack.LocationError.Starting -> {
                mapOf(KEY_TYPE to TYPE_LOCATION_ERROR_STARTING)
            }

            is HyperTrack.LocationError.Errors -> {
                mapOf(
                    KEY_TYPE to TYPE_LOCATION_ERROR_ERRORS,
                    KEY_VALUE to
                        locationError.errors
                            .map { serializeError(it) },
                )
            }
        }

    fun <T> parse(
        source: Serialized,
        parseFunction: (Parser) -> T,
    ): WrapperResult<T> {
        val parser = Parser(source)
        return try {
            if (parser.exceptions.isEmpty()) {
                Success(parseFunction.invoke(parser))
            } else {
                Failure(ParsingExceptions(source, parser.exceptions))
            }
        } catch (e: Exception) {
            Failure(
                if (parser.exceptions.isNotEmpty()) {
                    ParsingExceptions(source, parser.exceptions + e)
                } else {
                    e
                },
            )
        }
    }

    internal class Parser(
        private val source: Serialized,
    ) {
        private val _exceptions = mutableListOf<Exception>()
        val exceptions: List<Exception> = _exceptions

        inline fun <reified T> get(key: String): WrapperResult<T> =
            try {
                Success(source[key]!! as T)
            } catch (e: Exception) {
                Failure(
                    ParsingException(key, e)
                        .also {
                            _exceptions.add(it)
                        },
                )
            }

        inline fun <reified T> getOptional(key: String): WrapperResult<T?> =
            try {
                Success(source[key] as T?)
            } catch (e: Exception) {
                Failure(
                    ParsingException(key, e)
                        .also {
                            _exceptions.add(it)
                        },
                )
            }

        inline fun <reified T> assertValue(
            key: String,
            value: Any,
        ) {
            if (source[key] != value) {
                _exceptions.add(Exception("Assertion failed: $key != $value"))
            }
        }
    }

    internal data class ParsingExceptions(
        val source: Any,
        val exceptions: List<Exception>,
    ) : Throwable(
            exceptions
                .joinToString("\n")
                .let {
                    "Invalid input:\n\n${source}\n\n$it"
                },
        )

    internal class ParsingException(
        key: String,
        exception: Exception,
    ) : Exception("Invalid value for '$key': $exception", exception)

    private const val KEY_TYPE = "type"
    private const val KEY_VALUE = "value"

    private const val TYPE_RESULT_FAILURE = "failure"
    private const val TYPE_RESULT_SUCCESS = "success"

    private const val TYPE_ALLOW_MOCK_LOCATION = "allowMockLocation"
    private const val TYPE_DEVICE_ID = "deviceID"
    private const val TYPE_DYNAMIC_PUBLISHABLE_KEY = "dynamicPublishableKey"
    private const val TYPE_ERROR = "error"
    private const val TYPE_IS_AVAILABLE = "isAvailable"
    private const val TYPE_IS_INSIDE_GEOFENCE = "isInsideGeofence"
    private const val TYPE_IS_TRACKING = "isTracking"
    private const val TYPE_LOCATION = "location"
    private const val TYPE_LOCATION_WITH_DEVIATION = "locationWithDeviation"
    private const val TYPE_METADATA = "metadata"
    private const val TYPE_NAME = "name"
    private const val TYPE_ORDERS = "orders"
    private const val TYPE_WORKER_HANDLE = "workerHandle"

    private const val TYPE_LOCATION_ERROR_ERRORS = "errors"
    private const val TYPE_LOCATION_ERROR_NOT_RUNNING = "notRunning"
    private const val TYPE_LOCATION_ERROR_STARTING = "starting"

    private const val TYPE_ORDER_HANDLE = "orderHandle"
    private const val TYPE_GEOTAG_ORDER_STATUS_CLOCK_IN = "orderStatusClockIn"
    private const val TYPE_GEOTAG_ORDER_STATUS_CLOCK_OUT = "orderStatusClockOut"
    private const val TYPE_GEOTAG_ORDER_STATUS_CUSTOM = "orderStatusCustom"

    private const val KEY_LATITUDE = "latitude"
    private const val KEY_LONGITUDE = "longitude"

    private const val KEY_DEVIATION = "deviation"
    private const val KEY_GEOTAG_DATA = "data"
    private const val KEY_GEOTAG_EXPECTED_LOCATION = "expectedLocation"
    private const val KEY_GEOTAG_ORDER_HANDLE = "orderHandle"
    private const val KEY_GEOTAG_ORDER_STATUS = "orderStatus"
    private const val KEY_LOCATION = "location"
    private const val KEY_ORDER_HANDLE = "orderHandle"
    private const val KEY_ORDER_INDEX = "index"
}

File: sdk/ios/common/Serialization.swift
import HyperTrack

private let keyExpectedLocation = "expectedLocation"
private let keyGeotagData = "data"
private let keyType = "type"
private let keyValue = "value"

private let typeError = "error"
private let typeFailure = "failure"
private let typeIsAvailable = "isAvailable"
private let typeIsTracking = "isTracking"
private let typeLocation = "location"
private let typeMetadata = "metadata"
private let typeName = "name"
private let typeSuccess = "success"

func deserializeDynamicPublishableKey(
    _ args: [String: Any]
) -> Result<String, FailureResult> {
    if args[keyType] as? String != "dynamicPublishableKey" {
        return .failure(.fatalError(getParseError(args, key: keyType)))
    }
    guard let value = args[keyValue] as? String else {
        return .failure(.fatalError(getParseError(args, key: keyValue)))
    }
    return .success(value)
}

func deserializeGeotagData(
    _ args: [String: Any]
) -> Result<GeotagData, FailureResult> {
    guard let data = args[keyGeotagData] as? [String: Any] else {
        return .failure(.fatalError(getParseError(args, key: keyGeotagData)))
    }
    let orderHandleData = args["orderHandle"] as? [String: Any]
    let orderHandleResult: Result<HyperTrack.OrderHandle, FailureResult>? = if let orderHandleData = orderHandleData {
        deserializeOrderHandle(orderHandleData)
    } else {
        nil
    }
    if case let .failure(failure) = orderHandleResult {
        return .failure(failure)
    }
    let orderHandle: HyperTrack.OrderHandle? = if case let .success(orderHandle) = orderHandleResult {
        orderHandle
    } else {
        nil
    }

    let orderStatusData = args["orderStatus"] as? [String: Any]
    let orderStatusResult: Result<HyperTrack.OrderStatus, FailureResult>? = if let orderStatusData = orderStatusData {
        deserializeOrderStatus(orderStatusData)
    } else {
        nil
    }
    if case let .failure(failure) = orderStatusResult {
        return .failure(failure)
    }
    let orderStatus: HyperTrack.OrderStatus? = if case let .success(orderStatus) = orderStatusResult {
        orderStatus
    } else {
        nil
    }

    if let expectedLocationData = args[keyExpectedLocation] as? [String: Any] {
        let expectedLocation = deserializeLocation(expectedLocationData)
        switch expectedLocation {
        case let .failure(failure):
            return .failure(failure)
        case let .success(expectedLocation):
            return .success(.init(
                data: data,
                expectedLocation: expectedLocation,
                orderHandle: orderHandle,
                orderStatus: orderStatus
            ))
        }
    } else {
        return .success(.init(
            data: data,
            expectedLocation: nil,
            orderHandle: orderHandle,
            orderStatus: orderStatus
        ))
    }
}

func deserializeIsAvailable(_ data: [String: Any]) -> Result<Bool, FailureResult> {
    if data[keyType] as? String != typeIsAvailable {
        return .failure(.fatalError(getParseError(data, key: keyType)))
    }
    guard let value = data[keyValue] as? Bool else {
        return .failure(.fatalError(getParseError(data, key: keyValue)))
    }
    return .success(value)
}

func deserializeIsTracking(_ data: [String: Any]) -> Result<Bool, FailureResult> {
    if data[keyType] as? String != typeIsTracking {
        return .failure(.fatalError(getParseError(data, key: keyType)))
    }
    guard let value = data[keyValue] as? Bool else {
        return .failure(.fatalError(getParseError(data, key: keyValue)))
    }
    return .success(value)
}

func deserializeLocation(_ data: [String: Any]) -> Result<HyperTrack.Location, FailureResult> {
    if data[keyType] as? String != typeLocation {
        return .failure(.fatalError(getParseError(data, key: keyType)))
    }
    guard let value = data[keyValue] as? [String: Any] else {
        return .failure(.fatalError(getParseError(data, key: keyValue)))
    }
    guard let latitude = value["latitude"] as? Double else {
        return .failure(.fatalError(getParseError(value, key: "latitude")))
    }
    guard let longitude = value["longitude"] as? Double else {
        return .failure(.fatalError(getParseError(value, key: "longitude")))
    }
    return .success(.init(latitude: latitude, longitude: longitude))
}

func deserializeMetadata(_ data: [String: Any]) -> Result<[String: Any], FailureResult> {
    if data[keyType] as? String != typeMetadata {
        return .failure(.fatalError(getParseError(data, key: keyType)))
    }
    guard let value = data[keyValue] as? [String: Any] else {
        return .failure(.fatalError(getParseError(data, key: keyValue)))
    }
    return .success(value)
}

func deserializeName(_ data: [String: Any]) -> Result<String, FailureResult> {
    if data[keyType] as? String != typeName {
        return .failure(.fatalError(getParseError(data, key: keyType)))
    }
    guard let value = data[keyValue] as? String else {
        return .failure(.fatalError(getParseError(data, key: keyValue)))
    }
    return .success(value)
}

func deserializeOrderStatus(_ data: [String: Any]) -> Result<HyperTrack.OrderStatus, FailureResult> {
    guard let type = data[keyType] as? String else {
        return .failure(.fatalError(getParseError(data, key: keyType)))
    }
    switch type {
    case "orderStatusClockIn":
        return .success(.clockIn)
    case "orderStatusClockOut":
        return .success(.clockOut)
    case "orderStatusCustom":
        guard let value = data[keyValue] as? String else {
            return .failure(.fatalError(getParseError(data, key: keyValue)))
        }
        return .success(.custom(value))
    default:
        return .failure(.fatalError(getParseError(data, key: keyType)))
    }
}

func serializeDeviceID(_ deviceID: String) -> [String: Any] {
    return [
        keyType: "deviceID",
        keyValue: deviceID,
    ]
}

func serializeDynamicPublishableKey(_ dynamicPublishableKey: String) -> [String: Any] {
    return [
        keyType: "dynamicPublishableKey",
        keyValue: dynamicPublishableKey,
    ]
}

func serializeError(_ error: HyperTrack.Error) -> [String: Any] {
    let value: String
    switch error {
    case .blockedFromRunning:
        value = "blockedFromRunning"
    case .invalidPublishableKey:
        value = "invalidPublishableKey"
    case let .location(location):
        switch location {
        case .mocked:
            value = "location.mocked"
        case .servicesDisabled:
            value = "location.servicesDisabled"
        case .signalLost:
            value = "location.signalLost"
        }
    case let .permissions(permissionType):
        switch permissionType {
        case let .location(permission):
            switch permission {
            case .denied:
                value = "permissions.location.denied"
            case .insufficientForBackground:
                value = "permissions.location.insufficientForBackground"
            case .notDetermined:
                value = "permissions.location.notDetermined"
            case .provisional:
                value = "permissions.location.provisional"
            case .reducedAccuracy:
                value = "permissions.location.reducedAccuracy"
            case .restricted:
                value = "permissions.location.restricted"
            }
        }
    }

    return [
        keyType: typeError,
        keyValue: value,
    ]
}

func serializeIsAvailable(_ isAvailable: Bool) -> [String: Any] {
    return [
        keyType: typeIsAvailable,
        keyValue: isAvailable,
    ]
}

func serializeIsTracking(_ isTracking: Bool) -> [String: Any] {
    return [
        keyType: typeIsTracking,
        keyValue: isTracking,
    ]
}

func serializeLocateResult(_ locateResult: Result<HyperTrack.Location, Set<HyperTrack.Error>>) -> [String: Any] {
    switch locateResult {
    case let .success(success):
        return [
            keyType: typeSuccess,
            keyValue: serializeLocation(success),
        ]
    case let .failure(failure):
        return [
            keyType: typeFailure,
            keyValue: serializeErrors(failure),
        ]
    }
}

func serializeLocation(_ location: HyperTrack.Location) -> [String: Any] {
    return [
        keyType: typeLocation,
        keyValue: [
            "latitude": location.latitude,
            "longitude": location.longitude,
        ],
    ]
}

func serializeLocationError(_ error: HyperTrack.Location.Error) -> [String: Any] {
    switch error {
    case .starting:
        return [
            keyType: "starting",
        ]
    case .notRunning:
        return [
            keyType: "notRunning",
        ]
    case let .errors(errors):
        return [
            keyType: "errors",
            keyValue: serializeErrors(errors),
        ]
    }
}

func serializeLocationResult(_ result: Result<HyperTrack.Location, HyperTrack.Location.Error>) -> [String: Any] {
    switch result {
    case let .success(success):
        return [
            keyType: typeSuccess,
            keyValue: serializeLocation(success),
        ]
    case let .failure(failure):
        return [
            keyType: typeFailure,
            keyValue: serializeLocationError(failure),
        ]
    }
}

func serializeLocationWithDeviation(
    _ locationWithDeviation: HyperTrack.LocationWithDeviation
) -> [String: Any] {
    return [
        "location": serializeLocation(locationWithDeviation.location),
        "deviation": locationWithDeviation.deviation,
    ]
}

func serializeLocationWithDeviationResult(
    _ result: Result<HyperTrack.LocationWithDeviation, HyperTrack.Location.Error>
) -> [String: Any] {
    switch result {
    case let .success(success):
        return [
            keyType: typeSuccess,
            keyValue: [
                keyType: "locationWithDeviation",
                keyValue: serializeLocationWithDeviation(success),
            ],
        ]
    case let .failure(failure):
        return [
            keyType: typeFailure,
            keyValue: serializeLocationError(failure),
        ]
    }
}

func serializeMetadata(_ metadata: HyperTrack.JSON.Object) -> [String: Any] {
    return [
        keyType: typeMetadata,
        keyValue: toMap(metadata),
    ]
}

func serializeName(_ name: String) -> [String: Any] {
    return [
        keyType: typeName,
        keyValue: name,
    ]
}

func serializeErrors(_ errors: Set<HyperTrack.Error>) -> [[String: Any]] {
    return errors.map { error in
        serializeError(error)
    }
}

private func toMap(_ object: HyperTrack.JSON.Object) -> [String: Any] {
    return object.mapValues { value in
        switch value {
        case let .object(objectValue):
            return toMap(objectValue) as Any
        case let .array(arr):
            return toArray(arr)
        case let .bool(bool):
            return bool
        case let .number(number):
            return number
        case let .string(string):
            return string
        case .null:
            return NSNull()
        }
    }
}

private func toArray(_ array: [HyperTrack.JSON]) -> [Any] {
    return array.map { (json: HyperTrack.JSON) -> Any in
        switch json {
        case let .object(objectValue):
            return toMap(objectValue)
        case let .array(arr):
            return toArray(arr)
        case let .bool(bool):
            return bool
        case let .number(number):
            return number
        case let .string(string):
            return string
        case .null:
            return NSNull()
        }
    }
}

private func getParseError(_ data: Any, key: String) -> String {
    return "Invalid input for key \(key): \(data)"
}
